### PR TITLE
Account

### PR DESCRIPTION
**Pull request by a-nurmatov**

### CODE REVIEW

**Comment by a-nurmatov** [comment]
Hi, how are you doing?  Below lecturer gave these tasks as homework. If I am not mistaken I have already done the homework for this week in my last auth-page pull request review.

1. Separate logic from the main and login page into components
2. Add all static pages
3. Backlog work

File: client/src/app/account/components/account-card-list/account-card-list.component.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  @Output() addAccountClicked: EventEmitter<void> = new EventEmitter<void>();
+  constructor(
+    public dialog: MatDialog,
+    private accountService: AccountService
+  ) {
+    this.sendRequestToAccounts();
+    this.userAccounts$ = this.accountService.getUserAccounts();
+  }
+
+  ngOnInit(): void {}
```
Comment (AudriusBaltusis) [inline client/src/app/account/components/account-card-list/account-card-list.component.ts]: Please remove redundant `ngOnInit`

File: client/src/app/account/components/account-card-list/account-card-list.component.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    this.addAccountClicked.emit();
+  }
+
+  openDialogue(): void {
+    let dialogRef = this.dialog.open(DialogueComponent, {
+      height: '520px',
+      width: '600px',
+    });
+
+    dialogRef.afterClosed().subscribe((result) => {
```
Comment (AudriusBaltusis) [inline client/src/app/account/components/account-card-list/account-card-list.component.ts]: It looks redundant

File: client/src/app/account/components/account-card-list/account-card-list.component.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  sendRequestToAccounts(): void {
+    let userId = localStorage.getItem('userId');
+    if (userId) {
+      this.accountService.requestUserAccounts(userId);
+    }
+  }
+
+  cardTrackingFn(index: number, item: AccountInterface) {
```
Comment (AudriusBaltusis) [inline client/src/app/account/components/account-card-list/account-card-list.component.ts]: Please add return type

File: client/src/app/account/components/dialogue/dialogue.component.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  constructor(
+    private accountService: AccountService,
+    private snackBar: MatSnackBar
+  ) {}
+
+  ngOnInit(): void {
+    this.setUserDefaultCurrency();
+  }
+
+  onSubmit() {
```
Comment (AudriusBaltusis) [inline client/src/app/account/components/dialogue/dialogue.component.ts]: return type

File: client/src/app/account/components/dialogue/dialogue.component.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    const { title, currency, description } = this.addAccountForm.value;
+    const userId = localStorage.getItem('userId');
+    this.accountService
+      .addNewAccount(userId, title, currency, description)
+      .subscribe(
```
Comment (AudriusBaltusis) [inline client/src/app/account/components/dialogue/dialogue.component.ts]: You have to unsubscribe, or use `takeUntil()`

File: client/src/app/account/components/dialogue/dialogue.component.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    this.listOfCountries.forEach((country) => {
+      if (country.name === this.userCountry) {
+        this.userDefaultCurrency = country.currency;
+      }
+    });
+    this.addAccountForm.get('currency')!.setValue(this.userDefaultCurrency);
+  }
+
+  getTitleError(): string {
+    return this.addAccountForm.get('title')!.errors?.['required']
```
Comment (AudriusBaltusis) [inline client/src/app/account/components/dialogue/dialogue.component.ts:107]: I suggest to use `switch ... case` here

File: client/src/app/account/services/account.service.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import { Injectable } from '@angular/core';
+import { BehaviorSubject, Observable, Subject, tap } from 'rxjs';
+import { AccountInterface } from '../types/account.interface';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class AccountService {
+  accounts: AccountInterface[] = [];
+  accountsUpdated = new Subject<AccountInterface[]>();
```
Comment (AudriusBaltusis) [inline client/src/app/account/services/account.service.ts]: Seems it's private

File: client/src/app/account/services/account.service.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        {
+          userId,
+          title,
+          currency,
+          description,
+        }
+      )
+      .pipe(
+        tap(() => {
+          this.accounts.push({
```
Comment (AudriusBaltusis) [inline client/src/app/account/services/account.service.ts]: In this case better to call `requestUserAccounts` from component to keep account list up to date, than leave this responsibility to `addNewAccount`.

**Review by AudriusBaltusis — CHANGES_REQUESTED** [review]

File: client/src/app/account/components/dialogue/dialogue.component.ts (hunk 5 above)
Comment (a-nurmatov) [reply client/src/app/account/components/dialogue/dialogue.component.ts]: fixed

File: client/src/app/account/components/dialogue/dialogue.component.ts (hunk 6 above)
Comment (a-nurmatov) [reply client/src/app/account/components/dialogue/dialogue.component.ts:107]: get title errors giving object like this {required: true} or {pattern: true} checking with ternary operator or with if statement is better I think

**Comment by a-nurmatov** [comment]
Hi, how are you? 
I hope my code doesn't have serious problems

**Review by AudriusBaltusis — APPROVED** [review]

**Pull request merged**